### PR TITLE
definition / global context options

### PR DESCRIPTION
**Pull request by mirefek**
Update of the read_data API for definitions / clusters (not tested yet)

### CODE REVIEW

File: pytact/data_reader.py [hunk 1]
```diff
@@ -781,6 +771,16 @@ def is_file_representative(self) -> bool:
...
 
+    @classmethod
+    def _build_clusters(cls, definitions : Iterable[Definition]):
+        seen : set[Node] = set()
+        for definition in definitions:
+            node = definition.node
+            if node in seen: continue
+            cluster = list(definition.cluster)
+            yield cluster
+            seen.update(cluster)
```
Comment (LasseBlaauwbroek) [inline pytact/data_reader.py]: I don't think you really need this `seen` set. Couldn't you do something like this?
```suggestion
        for repr in definitions:
            yield list((repr, *itertools.takewhile(lambda d: d.cluster_representative == repr, definitions)))
```

File: pytact/data_reader.py [hunk 2]
```diff
@@ -568,53 +568,43 @@ def get_index(index):
...
-                seen.add(eprev.node)
-
-    @property
-    def global_context(self) -> Iterable[Definition]:
+    def global_context(self, across_files : bool = True, inclusive = False, seen : set[Node] = None) -> Iterable[Definition]:
         """All of the definitions in the global context when this definition was created.
 
         Note that this does not include this definition itself, except when the definition is a inductive,
         constructor or projection. Because those are mutually recursive objects, they reference themselves
         and are therefore part of their own global context.
```
Comment (LasseBlaauwbroek) [inline pytact/data_reader.py]: You should probably add some documentation for `across_files`, `inclusive` and `seen`. Also in the other functions that you modified with these parameters.

File: pytact/data_reader.py [hunk 3]
```diff
@@ -568,53 +568,43 @@ def get_index(index):
...
-        for eprev in self.external_previous:
-            if eprev.node not in seen:
-                cluster = list(eprev.cluster)
-                yield cluster
-                yield from cluster[-1]._clustered_global_context(seen)
-                seen.add(eprev.node)
-
-    @property
-    def global_context(self) -> Iterable[Definition]:
+    def global_context(self, across_files : bool = True, inclusive = False, seen : set[Node] = None) -> Iterable[Definition]:
```
Comment (LasseBlaauwbroek) [inline pytact/data_reader.py]: Do you have a particular usecase in mind to use `seen` from the public API? If not, maybe this should be an auxiliary function and the real function doesn't have this argument?

File: pytact/data_reader.py [hunk 4]
```diff
@@ -568,53 +568,43 @@ def get_index(index):
...
         """All of the definitions in the global context when this definition was created, clustered into
         mutually recursive cliques.
 
         The definition itself may be part of the first mutually recursive cluster."""
-        start_cluster = list(self.cluster)
-        match start_cluster[0].kind:
-            case Definition.Inductive(_) | Definition.Constructor(_) | Definition.Projection(_):
-                yield list(start_cluster)
-        yield from start_cluster[-1]._clustered_global_context(set())
+        return self._build_clusters(self.global_context(
```
Comment (LasseBlaauwbroek) [inline pytact/data_reader.py]: Maybe call it `_split_by_clusters`?

**Review by LasseBlaauwbroek — CHANGES_REQUESTED** [review]
Some comments. Otherwise this looks good!

File: pytact/data_reader.py (hunk 1 above)
Comment (LasseBlaauwbroek) [reply pytact/data_reader.py]: Well, I guess you cannot use this when the iterable is coming from `Dataset.definitions(spine_only=False)`. But maybe you should have a special case for that in `Dataset`? It feels like this logic should not be part of `Definition`.

Comment (mirefek) [reply pytact/data_reader.py]: I don't see a problem when the iterator is a sequence

Comment (mirefek) [reply pytact/data_reader.py]: I was not sure if I can count on the clusters being grouped together, fine, I will use something like that (after I understand your magical code :-) )

Comment (mirefek) [reply pytact/data_reader.py]: itertools.groupby could be useful

Comment (LasseBlaauwbroek) [reply pytact/data_reader.py]: Well, the problem is mainly that I'm not sure if the list of definitions is encoded in the proper ordering such that
1. Clusters occur next to each other.
2. The representative of a cluster occurs first.
But I guess I can check if that holds.

Also, when you iterate over a Sequence, do you actually consume it? Not sure.

File: pytact/data_reader.py (hunk 3 above)
Comment (mirefek) [reply pytact/data_reader.py]: You are right, I just don't like auxiliary functions much :-), but you are probably right, let's make an auxiliary function with `seen`

File: pytact/data_reader.py (hunk 4 above)
Comment (mirefek) [reply pytact/data_reader.py]: why not, I didn't know a really good fitting name for it

Comment (mirefek) [reply pytact/data_reader.py]: `group_by_clusters` ?

File: pytact/data_reader.py (hunk 2 above)
Comment (mirefek) [reply pytact/data_reader.py]: OK

File: pytact/data_reader.py (hunk 4 above)
Comment (LasseBlaauwbroek) [reply pytact/data_reader.py]: Sounds good.

File: pytact/data_reader.py (hunk 1 above)
Comment (mirefek) [reply pytact/data_reader.py]: When using `itertools.group_by`, (2) is not necessary. About (1), I don't know, it is a question about the dataset.
Ad iterating over a sequence, yes, it should work the same way as iterating over a list -- it gets automatically converted into an iterator.

Comment (LasseBlaauwbroek) [reply pytact/data_reader.py]: I think that the invariants I mentioned above hold even for `Dataset.defninitions`. But I'll check more after I finish my meeting with Jason.

File: pytact/data_reader.py [hunk 5]
```diff
@@ -781,6 +788,16 @@ def is_file_representative(self) -> bool:
         Se also `Dataset.representative`."""
         return self._lreader[self._graph].representative == self.node.nodeid
 
+    @classmethod
+    def _group_by_clusters(cls, definitions : Iterable[Definition]) -> Iterable[list[Definition]]:
```
Comment (LasseBlaauwbroek) [inline pytact/data_reader.py:793]:
```suggestion
    def _group_by_clusters(cls, definitions : Iterable[Definition]) -> Iterable[Iterable[Definition]]:
```
I think that with the new code, the type is no longer correct?

File: pytact/data_reader.py [hunk 6]
```diff
@@ -818,83 +835,58 @@ def representative(self) -> Definition | None:
...
+            graph = self._graph
+            lreader = self._lreader
+            ds = self._reader.definitions
+            count = len(ds)
+            return TupleLike(
+                count,
+                lambda index: cast(Definition, Node(graph, ds[index], lreader).definition)
+            )
+
+    def clustered_definitions(self, across_files = False, spine_only = False) -> Iterable[list[Definition]]:
```
Comment (LasseBlaauwbroek) [inline pytact/data_reader.py:876]:
```suggestion
    def clustered_definitions(self, across_files = False, spine_only = False) -> Iterable[Iterable[Definition]]:
```

**Review by LasseBlaauwbroek — CHANGES_REQUESTED** [review]
I checked the dataset, and it seems that it does indeed satisfy the conditions needed to make this code work. It's a bit fragile maybe, but for now I think it's fine.

I added two more suggestions, but otherwise this looks good.

File: pytact/data_reader.py (hunk 5 above)
Comment (mirefek) [reply pytact/data_reader.py:793]: to keep the original type, there is the `lambda x: list(x[1])`, I could keep it as an only iterator with replacing it by `lambda x: x[1]`

Comment (LasseBlaauwbroek) [reply pytact/data_reader.py:793]: Oh right, I missed that. I think an `Iterable[Iterable[Definition]]` is better. People can always cast it to a list.

Comment (mirefek) [reply pytact/data_reader.py:793]: ok, I will change it, and test the code tomorrow (I expect some bugs, it is not tested so far)

Comment (mirefek) [reply pytact/data_reader.py:793]: I changed it, tested it, and consequently changed it back :-) -- without the packing into list, itertools.groupby is fragile

**Review by LasseBlaauwbroek — APPROVED** [review]
Okay, everything looks good to me. Is this ready to merge?

**Comment by LasseBlaauwbroek** [comment]
@mirefek , can I get a reply on this. I'd like to merge.

**Comment by mirefek** [comment]
Oh sorry, sure, ready to merge

**Comment by mirefek** [comment]
we could still find further bugs while testing the new dataserver but I think it was good to merge for now

**Pull request merged**